runtime/lua/vim/filetype.lua

```diff
@@ -2407,39 +2505,54 @@ local function dispatch(ft, path, bufnr, ...)
   return ft0, on_detect
 end
 
---- Lookup table/cache for patterns that contain an environment variable pattern, e.g. ${SOME_VAR}.
---- @type table<string,boolean>
-local expand_env_lookup = {}
+--- Lookup table/cache for patterns
+--- @alias vim.filetype.pattern_cache { fullpat: string, has_env: boolean, has_slash: boolean }
+--- @type table<string,vim.filetype.pattern_cache>
+local pattern_lookup = {}
+
+--- @param pat string
+--- @return vim.filetype.pattern_cache
+local function parse_pattern(pat)
+  pattern_lookup[pat] = {
+    fullpat = '^' .. pat .. '$',
+    has_env = pat:find('%$%b{}') ~= nil,
+    has_slash = pat:find('/') ~= nil,
+  }
+  return pattern_lookup[pat]
+end
+
+--- @param pat string
+--- @return boolean
+--- @return string
+local function expand_envvar_pattern(pat)
+  local some_env_missing = false
+  local expanded = pat:gsub('%${(%S-)}', function(env)
+    local val = vim.env[env] --- @type string?
+    some_env_missing = some_env_missing or val == nil
+    return vim.pesc(val or '')
+  end)
+  return some_env_missing, expanded
+end
 
 --- @param name string
 --- @param path string
 --- @param tail string
 --- @param pat string
---- @return string|false?
+--- @return string|boolean?
 local function match_pattern(name, path, tail, pat)
-  if expand_env_lookup[pat] == nil then
-    expand_env_lookup[pat] = pat:find('%${') ~= nil
-  end
-  if expand_env_lookup[pat] then
-    local return_early --- @type true?
-    --- @type string
-    pat = pat:gsub('%${(%S-)}', function(env)
-      -- If an environment variable is present in the pattern but not set, there is no match
-      if not vim.env[env] then
-        return_early = true
-        return nil
-      end
-      return vim.pesc(vim.env[env])
-    end)
-    if return_early then
+  local pat_cache = pattern_lookup[pat] or parse_pattern(pat)
```

Comment:
Can we move the pattern parsing to *add* time as opposed to *match* time?

If this has to be deferred until the first autocmd is triggered, then that can be done.

Reply:
It can be moved to `sort_by_priority()`, as it does some inferring inside of it. However, I am not sure what the benefit of this would be. Could you elaborate, please?

Reply:
*match* time is hotter than *add* time. I think it's best we do as much pre-processing at *add* time, and keep matching as simple as possible.

Reply:
This is resolved in #29632.